File: 2021/star-14.py
import sys
import pprint

def sequence(max):
    ret = [];

    current = 0;
    for i in xrange(0, max + 1):
        current = current + i
        ret.append(current)

    return ret


def fuel_required(pos, crabs, costs):
    sum = 0
    for crab in crabs:
        move = abs(crab - pos)
        sum = sum + costs[move]

    return sum


def run(filename, verbose):
    lines = []

    with open(filename) as file:
        lines = [line.strip() for line in file.readlines()]

    positions = [int(v) for v in lines[0].split(',')]
    if (verbose):
        pprint.pprint(positions)

    max = positions[0]
    min = positions[0]

    for p in positions:
        max = p if p > max else max
        min = p if p < min else min

    costs = sequence(max)
    if (verbose):
        cost_str = ",".join([ str(c) for c in costs ])
        print("costs: " + cost_str)

    best = fuel_required(min, positions, costs)
    for i in xrange(min, max + 1):
        fuel = fuel_required(i, positions, costs)
        if (verbose):
            print('Fuel required for pos %d: %d' % (i, fuel))

        if (fuel < best):
            best = fuel
            if (verbose):
                print('Best so far is position %d with a cost of %d' % (i, fuel))

    return best

expected = 168;
actual = run('star-13-example.txt', True);
if actual == expected:
    print("Example answer OK: %d." % (actual));
else:
    print('You broke the example, expected: %d, actual: %d.' % (expected, actual))
    sys.exit()


output = run('star-13-input.txt', False);

print("The puzzle answer is: %d" % (output));

File: 2021/star-14.php
<?php

function sequence($max)
{
    $ret = [];

    $current = 0;
    for ($i = 0; $i <= $max; $i++) {
        $current += $i;
        $ret[] = $current;
    }

    return $ret;
}

function fuel_required($pos, $crabs, $costs)
{
    $sum = 0;
    foreach ($crabs as $crab) {
        $sum += $costs[abs($crab - $pos)];
    }

    return $sum;
}

function run($filename, $verbose = false)
{
    $lines = array_map('trim', file($filename));

    $positions = explode(',', array_shift($lines));
    $max = $positions[0];
    $min = $positions[0];

    foreach ($positions as $pos) {
        $max = $pos > $max ? $pos : $max;
        $min = $pos < $min ? $pos : $min;
    }

    $costs = sequence($max);
    if ($verbose) {
        echo 'costs: ' . implode(',', $costs) . "\n";
    }

    $best = fuel_required($min, $positions, $costs);

    for ($i = $min; $i <= $max; $i++) {
        $fuel = fuel_required($i, $positions, $costs);
        if ($verbose) {
            echo "Fuel required for pos $i: $fuel\n";
        }

        if ($fuel < $best) {
            $best = $fuel;
            if ($verbose) {
                echo "Best so far is position $i with a cost of $fuel\n";
            }
        }
    }

    return $best;
}

$expected = 168;
$actual = run('star-13-example.txt', true);
if ($actual !== $expected) {
    echo 'You broke the example, ' . "expected: $expected, actual: $actual.\n";
    die();
} else {
    echo "Example answer OK: $actual\n";
}

$output = run('star-13-input.txt');

echo "The puzzle answer is:  $output\n";
